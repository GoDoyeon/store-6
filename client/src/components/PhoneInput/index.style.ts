import styled from 'styled-components';
import { phoneUnderLineSVG } from '~/assets';

export const PhoneWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  bottom: 4px;
`;

export const StyledPhoneInput = styled.input`
  text-align: center;
  box-sizing: border-box;
  width: 100px;
  height: 34px;
  padding: 0 16px 10px 16px;
  margin-top: 16px;
  line-height: 25px;
  font-size: 14px;
  background: bottom left / contain no-repeat url(${phoneUnderLineSVG});

  &::placeholder {
    color: #999;
  }
`;
