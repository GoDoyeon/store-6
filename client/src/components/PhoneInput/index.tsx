import React, { Fragment, useCallback, useMemo } from 'react';
import Input from '~/components/Input';
import Divider from '~/components/Divider';

import { PhoneWrapper } from './index.style';
import { REG_NOT_DIGITS } from '~/utils/validation';

interface Props {
  onChange?: (phoneNumber: string) => void;
  value?: string;
  disabled?: boolean;
  placeholder?: string;
}

const phoneNumbersInfo = [
  { key: 1, max: 3, placeholder: '입력이' },
  { key: 2, max: 4, placeholder: '필요' },
  { key: 3, max: 4, placeholder: '합니다' },
];

const PhoneInput: React.FC<Props> = ({
  onChange,
  value,
  disabled,
  placeholder,
}) => {
  const phoneValue = useMemo(() => {
    const result = phoneNumbersInfo.map(() => '');
    value?.split('-').forEach((v, i) => {
      result[i] = v;
    });
    return result;
  }, [value]);
  const phonePlaceHolder = useMemo(() => {
    const result = phoneNumbersInfo.map((info) => info.placeholder);
    placeholder
      ?.split(' ')
      .slice(0, 3)
      .forEach((p, i) => {
        result[i] = p;
      });
    return result;
  }, [placeholder]);

  const setPhoneNumber = useCallback(
    (i: number) => (e: React.ChangeEvent<HTMLInputElement>) => {
      const nextPhone = [...phoneValue];
      const currentValue = e.target.value;
      const phoneNumber = currentValue.replace(REG_NOT_DIGITS, '');

      if (phoneNumber === phoneValue[i]) return;
      nextPhone[i] = phoneNumber;
      onChange(nextPhone.join('-'));
    },
    [onChange, phoneValue],
  );

  const setPhoneNumbers = useMemo(
    () => phoneNumbersInfo.map((_, i) => setPhoneNumber(i)),
    [setPhoneNumber],
  );

  return (
    <PhoneWrapper>
      {phoneNumbersInfo.map((phoneNumberInfo, i) => {
        return (
          <Fragment key={phoneNumberInfo.key}>
            <Input
              type="text"
              width="90px"
              value={phoneValue[i]?.slice(0, phoneNumberInfo.max) || ''}
              onChange={setPhoneNumbers[i]}
              maxLength={phoneNumberInfo.max}
              style={{ textAlign: 'center', fontSize: '14px' }}
              disabled={disabled}
              placeholder={phonePlaceHolder[i]}
            />
            {i !== phoneNumbersInfo.length - 1 && (
              <Divider
                style={{ margin: '20px' }}
                width="10px"
                direction="horizontal"
              />
            )}
          </Fragment>
        );
      })}
    </PhoneWrapper>
  );
};

export default PhoneInput;
