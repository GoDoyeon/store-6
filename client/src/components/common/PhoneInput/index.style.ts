import styled from 'styled-components';
import { phoneUnderLineSVG } from '~/assets';

export const PhoneWrapper = styled.div<{ fontSize }>`
  display: flex;
  align-items: center;
  justify-content: space-between;
  position: relative;
  font-size: ${({ fontSize }) => fontSize || 'inherit'};
`;

export const StyledPhoneInput = styled.input`
  text-align: center;
  box-sizing: border-box;
  width: 100px;
  height: 34px;
  background: bottom left / contain no-repeat url(${phoneUnderLineSVG});
  font-size: inherit;
  line-height: inherit;

  &::placeholder {
    color: #999;
  }
`;
