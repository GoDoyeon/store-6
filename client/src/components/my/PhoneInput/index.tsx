import React, { Fragment, useCallback, useMemo } from 'react';

import { PhoneWrapper, StyledPhoneInput } from './index.style';
import { REG_NOT_DIGITS } from '~/utils/validation';
import { hyphenSVG } from '~/assets';

interface Props {
  onChange?: (phoneNumber: string) => void;
  value?: string;
  disabled?: boolean;
  placeholder?: string;
}

const phoneNumbersInfo = [
  { key: 1, max: 3, placeholder: '입력이' },
  { key: 2, max: 4, placeholder: '필요' },
  { key: 3, max: 4, placeholder: '합니다' },
];

const PhoneInput: React.FC<Props> = ({
  onChange,
  value,
  disabled,
  placeholder,
}) => {
  const phoneValue = useMemo(() => {
    const result = phoneNumbersInfo.map(() => '');
    value?.split('-').forEach((v, i) => {
      result[i] = v;
    });
    return result;
  }, [value]);
  const phonePlaceHolder = useMemo(() => {
    const result = phoneNumbersInfo.map((info) => info.placeholder);
    placeholder
      ?.split(' ')
      .slice(0, 3)
      .forEach((p, i) => {
        result[i] = p;
      });
    return result;
  }, [placeholder]);

  const setPhoneNumber = useCallback(
    (i: number) => (e: React.ChangeEvent<HTMLInputElement>) => {
      const nextPhone = [...phoneValue];
      const currentValue = e.target.value;
      const phoneNumber = currentValue.replace(REG_NOT_DIGITS, '');

      if (phoneNumber === phoneValue[i]) return;
      nextPhone[i] = phoneNumber;
      onChange(nextPhone.join('-'));
    },
    [onChange, phoneValue],
  );

  const setPhoneNumbers = useMemo(
    () => phoneNumbersInfo.map((_, i) => setPhoneNumber(i)),
    [setPhoneNumber],
  );

  return (
    <PhoneWrapper>
      {phoneNumbersInfo.map((phoneNumberInfo, i) => {
        return (
          <Fragment key={phoneNumberInfo.key}>
            <StyledPhoneInput
              type="text"
              value={phoneValue[i]?.slice(0, phoneNumberInfo.max) || undefined}
              onChange={setPhoneNumbers[i]}
              maxLength={phoneNumberInfo.max}
              disabled={disabled}
              placeholder={phonePlaceHolder[i]}
            />
            {i !== phoneNumbersInfo.length - 1 && (
              <img style={{ margin: '10px' }} src={hyphenSVG} alt="hyphen" />
            )}
          </Fragment>
        );
      })}
    </PhoneWrapper>
  );
};

export default PhoneInput;
