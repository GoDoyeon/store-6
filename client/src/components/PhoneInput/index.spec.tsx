import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import PhoneInput from './index';

describe('<PhoneInput />', () => {
  it('should render same with snapshot', () => {
    const { container } = render(<PhoneInput />);
    expect(container).toMatchSnapshot();
  });
});
