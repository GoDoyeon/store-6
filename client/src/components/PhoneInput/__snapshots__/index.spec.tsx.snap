// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PhoneInput /> should render same with snapshot 1`] = `
.c1 {
  box-sizing: border-box;
  width: 90px;
  height: 34px;
  padding: 0 16px 10px 16px;
  margin-top: 16px;
  line-height: 25px;
  font-size: 20px;
  background: bottom left / contain no-repeat url();
}

.c1::-webkit-input-placeholder {
  color: #999;
}

.c1::-moz-placeholder {
  color: #999;
}

.c1:-ms-input-placeholder {
  color: #999;
}

.c1::placeholder {
  color: #999;
}

.c2 {
  width: 10px;
  height: 1px;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  background: url(https:/store-6-bucket.s3.ap-northeast-2.amazonaws.com/common/line.svg) bottom left no-repeat;
  background-size: cover;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  bottom: 4px;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      maxlength="3"
      placeholder="입력이"
      style="text-align: center; font-size: 14px;"
      type="text"
      value=""
      width="90px"
    />
    <div
      class="c2"
      direction="horizontal"
      style="margin: 20px;"
      width="10px"
    />
    <input
      class="c1"
      maxlength="4"
      placeholder="필요"
      style="text-align: center; font-size: 14px;"
      type="text"
      value=""
      width="90px"
    />
    <div
      class="c2"
      direction="horizontal"
      style="margin: 20px;"
      width="10px"
    />
    <input
      class="c1"
      maxlength="4"
      placeholder="합니다"
      style="text-align: center; font-size: 14px;"
      type="text"
      value=""
      width="90px"
    />
  </div>
</div>
`;
